app/validators/answers/answer_validator.py

```diff
@@ -1,11 +1,122 @@
 from app.answer_type import AnswerType
+from app.validators.questionnaire_schema import get_object_containing_key
 from app.validators.validator import Validator
 
 
 class AnswerValidator(Validator):
-    def __init__(self, schema_element):
+    OPTION_MISSING_Q_CODE = "Option q_code must be provided"
+    ANSWER_MISSING_Q_CODE = "Answer q_code must be provided"
+    NON_CHECKBOX_OPTION_HAS_Q_CODE = "Non checkbox option cannot contain q_code"
+    DETAIL_ANSWER_MISSING_Q_CODE = "Detail answer q_code must be provided"
+    CHECKBOX_DETAIL_ANSWER_HAS_Q_CODE = "Checkbox detail answer cannot contain q_code"
+    CONFIRMATION_QUESTION_HAS_Q_CODE = "Confirmation question cannot contain q_code"
+    DATA_VERSION_NOT_0_0_1_Q_CODE_PRESENT = (
+        "q_code can only be used with data_version 0.0.1"
+    )
+    CHECKBOX_ANSWER_AND_OPTIONS_Q_CODE_MUTUALLY_EXLUSIVE = (
+        "Checkbox answer and option q_code are mutually exclusive"
+    )
+    CHECKBOX_ANSWER_OR_OPTIONS_MUST_HAVE_Q_CODES = (
+        "Either checkbox answer or options must have q_codes"
+    )
+
+    def __init__(self, schema_element, questionnaire_schema):
         super().__init__(schema_element)
         self.answer = schema_element
         self.answer_id = self.answer["id"]
         self.answer_type = AnswerType(self.answer["type"])
         self.context["answer_id"] = self.answer_id
+        self.questionnaire_schema = questionnaire_schema
+
+    def validate(self):
+        self._validate_q_codes()
+
+        return self.errors
+
+    def _validate_q_codes(self):
+        is_confirmation_question = (
+            self.questionnaire_schema.get_block_by_answer_id(self.answer_id).get("type")
+            == "ConfirmationQuestion"
+        )
+
+        if (
+            self.questionnaire_schema.schema["data_version"] != "0.0.1"
+            or is_confirmation_question
+        ):
+            has_q_code = get_object_containing_key(self.answer, key_name="q_code")
+            if has_q_code:
+                self.add_error(
+                    self.CONFIRMATION_QUESTION_HAS_Q_CODE
+                    if is_confirmation_question
+                    else self.DATA_VERSION_NOT_0_0_1_Q_CODE_PRESENT,
+                    answer_id=self.answer["id"],
+                )
+
+            return None
+
+        if self.answer_type == AnswerType.CHECKBOX:
```

Comment:
Minor, these should actually be identity checks instead of equality.
```suggestion
        if self.answer_type is AnswerType.CHECKBOX:
```